curvefs/src/mds/topology/topology_manager.cpp

```diff
@@ -0,0 +1,855 @@
+                       << " failed. fsId = " << fsId
+                       << ", poolId = " << poolId
+                       << ", copysetId = " << copysetId
+                       << ", partitionId = " << partitionId;
+            return kTopoErrCodeCreatePartitionFailed;
+        }
+    }
+    return kTopoErrCodeSuccess;
+}
+
+int TopologyManager::CreateCopyset() {
+    PoolIdType poolId;
+    int ret = topology_->ChooseSinglePoolRandom(&poolId);
+    if (kTopoErrCodeSuccess != ret) {
+        LOG(ERROR) << "Select Pool failed when create partition"
+                   << "erro code = " << ret;
+        return ret;
+    }
+
+    Pool pool;
+    if (!topology_->GetPool(poolId, &pool)) {
+        LOG(ERROR) << "Get pool failed.";
+        return kTopoErrCodePoolNotFound;
+    }
+
+    int replicaNum = pool.GetReplicaNum();
+    std::set<ZoneIdType> zones;
+    ret = topology_->ChooseZonesInPool(poolId, &zones, replicaNum);
+    if (kTopoErrCodeSuccess != ret) {
+        LOG(ERROR) << "Select Zone failed when create partition"
+                   << "poolId = " << poolId
+                   << "erro code = " << ret;
+        return ret;
+    }
+
+    std::set<MetaServerIdType> metaServerIds;
+    std::set<std::string> metaServerAddrs;
+    for (auto zoneId : zones) {
+        MetaServerIdType metaServerId;
+        ret = topology_->ChooseSingleMetaServerInZone(zoneId, &metaServerId);
+        if (kTopoErrCodeSuccess != ret) {
+            LOG(ERROR) << "Select metaServer failed when create partition"
+                       << "zoneId = " << zoneId
+                       << "erro code = " << ret;
+            return ret;
+        }
+        MetaServer metaServer;
+        if (topology_->GetMetaServer(metaServerId, &metaServer)) {
+            metaServerAddrs.emplace(BuildPeerId(metaServer.GetInternalHostIp(),
+                                                metaServer.GetInternalPort()));
+        } else {
+            LOG(ERROR) << "Get metaserver failed.";
+            return kTopoErrCodeMetaServerNotFound;
+        }
+        metaServerIds.emplace(metaServerId);
+    }
+
+    if (metaServerAddrs.size() != replicaNum) {
+        LOG(ERROR) << "Have invalid metaserver number when create copyset."
+                   << " metaserver number = " << metaServerAddrs.size()
+                   << ", replicanum = " << replicaNum;
+    }
+
+    // send create copyset request
+    CopySetIdType copysetId = topology_->AllocateCopySetId(poolId);
+    FSStatusCode retcode = metaserverClient_->CreateCopySet(poolId, copysetId,
+                                                            metaServerAddrs);
+    CopySetInfo copyset(poolId, copysetId);
+    copyset.SetCopySetMembers(metaServerIds);
+    if (FSStatusCode::OK == retcode) {
+        int ret = topology_->AddCopySet(copyset);
+        if (kTopoErrCodeSuccess != ret) {
+            LOG(ERROR) << "Add copyset failed after create copyset."
```

Comment:
delete copyset from metaserver

Reply:
> delete copyset from metaserver

This will involve some error handling, add a TODO first. (delete copyset interface does not exist now)